src/salame/Salame.cpp

```diff
@@ -12,84 +12,130 @@
+        }
 
         // STEP 2: Calculate the contribution to Ez from only the SALAME beam
 
-        // deposit SALAME beam jz
-        hipace->m_multi_beam.DepositCurrentSlice(hipace->m_fields, hipace->m_3D_geom, lev, step,
-            islice_local, false, true, false, WhichSlice::Salame);
+        for (int lev=0; lev<current_N_level; ++lev) {
+            // deposit SALAME beam jz
+            hipace->m_multi_beam.DepositCurrentSlice(hipace->m_fields, hipace->m_3D_geom, lev, step,
+                islice_local, false, true, false, WhichSlice::Salame);
+        }
 
-        SalameInitializeSxSyWithBeam(hipace, lev);
+        for (int lev=0; lev<current_N_level; ++lev) {
+            SalameInitializeSxSyWithBeam(hipace, lev);
+        }
 
-        hipace->ExplicitMGSolveBxBy(lev, WhichSlice::Salame);
+        for (int lev=0; lev<current_N_level; ++lev) {
+            hipace->ExplicitMGSolveBxBy(lev, WhichSlice::Salame);
+        }
 
-        hipace->m_fields.setVal(0., lev, WhichSlice::Salame, "Ez", "jx", "jy");
+        for (int lev=0; lev<current_N_level; ++lev) {
+            hipace->m_fields.setVal(0., lev, WhichSlice::Salame, "Ez", "jx", "jy");
+        }
 
         // get jx jy (SALAME only) on the next slice using Bx By (SALAME only) on this slice
         if (do_advance) {
-            SalameOnlyAdvancePlasma(hipace, lev);
+            if (hipace->m_N_level > 1) {
+                // tag to prev slice for ux uy push
+                hipace->m_multi_plasma.TagByLevel(current_N_level, hipace->m_3D_geom, true);
+            }
 
-            hipace->m_multi_plasma.DepositCurrent(hipace->m_fields, hipace->m_multi_laser,
+            for (int lev=0; lev<current_N_level; ++lev) {
+                SalameOnlyAdvancePlasma(hipace, lev);
+            }
+
+            if (hipace->m_N_level > 1) {
+                // tag to temp slice for deposition
+                hipace->m_multi_plasma.TagByLevel(current_N_level, hipace->m_3D_geom);
+            }
+
+            for (int lev=0; lev<current_N_level; ++lev) {
+                hipace->m_multi_plasma.DepositCurrent(hipace->m_fields, hipace->m_multi_laser,
                     WhichSlice::Salame, true, false, false, false, false, hipace->m_3D_geom, lev);
+            }
         } else {
-            SalameGetJxJyFromBxBy(hipace, lev);
+            for (int lev=0; lev<current_N_level; ++lev) {
+                SalameGetJxJyFromBxBy(hipace, lev);
+            }
         }
 
-        hipace->m_fields.SolvePoissonEz(hipace->m_3D_geom, lev, WhichSlice::Salame);
+        hipace->m_fields.SolvePoissonEz(hipace->m_3D_geom, current_N_level, WhichSlice::Salame);
 
         // STEP 3: find ideal weighting factor of the SALAME beam using the computed Ez fields,
         // and update the beam with it
 
+        for (int lev=0; lev<current_N_level; ++lev) {
+            hipace->m_fields.setVal(0., lev, WhichSlice::Salame, "jz_beam");
+            // deposit SALAME beam jz only on the highest level for SalameGetW
```

Comment:
```suggestion
            // deposit SALAME beam jz only on the highest level for SalameGetW,
            // since the most accurate field (on the highest level) is supposed to be flattened
```
Just a bit of clarification